envoy/stats/stats.h

```diff
@@ -92,6 +92,11 @@ class Metric : public RefcountInterface {
    */
   virtual bool used() const PURE;
 
+  /**
+   * Indicates whether this metric is used or check in the ForEach loops
```

Comment:
We should say more about exactly what APIs this affects. I was thinking maybe we *would* want to hit this in forEach loops but we would want to leave it up to the stats sink and admin panel to skip them.

Note that for stats-sinks there's already a sink-predicate concept this could be an extension of that.

I'm wondering also if it might be interesting for our own devs relying on this feature to be able to look at the hidden stats from the admin panel, say with a `/stats?hidden` parameter.

Reply:
I was thinking it's cleaner to have forEach hit all the data structures cleanly; the caller supplying the functor can then apply used() and hidden() (I slightly prefer 'hidden' to 'internal') per policy they choose.

Reply:
That sounds reasonable to me as well, especially if there is precedent. It would likely also mean that we won't need to do anything special for the hot restart case!

Reply:
I reverted the change in the ForEach loops